src/python/pants/backend/python/tasks2/pytest_run.py

```diff
@@ -550,16 +557,137 @@ def _run_tests(self, targets):
+      safe_mkdir(coverage_workdir)
+      return coverage_workdir
+
+    def files(self):
+      def files_iter():
+        for dir_path, _, file_names in os.walk(self.root_dir):
+          for filename in file_names:
+            yield os.path.join(dir_path, filename)
+      return list(files_iter())
+
+  # TODO(John Sirois): Its probably worth generalizing a means to mark certain options or target
+  # attributes as making results un-cacheable. See: https://github.com/pantsbuild/pants/issues/4748
+  class NeverCacheFingerprintStrategy(DefaultFingerprintStrategy):
+    def compute_fingerprint(self, target):
+      return uuid.uuid4()
+
+  def _fingerprint_strategy(self):
+    if self.get_options().profile:
+      # A profile is machine-specific and we assume anyone wanting a profile wants to run it here
+      # and now and not accept some old result, even if on the same inputs.
+      return self.NeverCacheFingerprintStrategy()
+    else:
+      return None  # Accept the default fingerprint strategy.
+
+  def _do_run_tests(self, partition):
+    with self.invalidated(partition,
+                          fingerprint_strategy=self._fingerprint_strategy(),
                           # Re-run tests when the code they test (and depend on) changes.
                           invalidate_dependents=True) as invalidation_check:
 
-      invalid_tgts = [tgt for vts in invalidation_check.invalid_vts for tgt in vts.targets]
-      result = self._run_pytest(invalid_tgts)
-      return result.checked()
+      invalid_tgts = [invalid_tgt
+                      for vts in invalidation_check.invalid_vts
+                      for invalid_tgt in vts.targets]
+
+      # Processing proceeds through:
+      # 1.) output -> workdir
+      # 2.) [iff all == invalid] workdir -> cache: We do this manually for now.
+      # 3.) [iff invalid == 0 and all > 0] cache -> workdir: Done transparently by `invalidated`.
+      # 4.) [iff finals] workdir -> finals: We perform this step as an unconditional post-process.
+
+      # 1.) Write all results that will be potentially cached to workdir.
+      workdirs = self.Workdirs.for_targets(self.workdir, partition)
+      result = self._run_pytest(workdirs, invalid_tgts).checked()
+
+      cache_vts = self._vts_for_partition(invalidation_check)
+      if invalidation_check.all_vts == invalidation_check.invalid_vts:
+        # 2.) The full partition was invalid, cache results.
+        if result.success and self.artifact_cache_writes_enabled():
+          self.update_artifact_cache([(cache_vts, workdirs.files())])
+      elif not invalidation_check.invalid_vts:
+        # 3.) The full partition was valid, our results will have been staged for/by caching if not
+        # already local.
+        pass
+      else:
+        # The partition was partially invalid.
 
-  def _run_pytest(self, targets):
+        # We don't cache results; so others will need to re-run this partition.
```

Comment:
I'm a bit confused by this (could be my oxygen-starved brain on this flight), but why not cache here? Isn't the workdir supposed to be in the exact same state, post test-running, regardless of whether all or only some of the targets were invalid?